Adding MS SQL translation

edgararuiz opened the pull request:
New translation allows dplyr to select, group, summarise and join

imanuelcostigan commented:
@edgararuiz have you looked at [RSQLServer](https://github.com/imanuelcostigan/RSQLServer)? I incorporated a dplyr backend in this package because dplyr wasn't going to include new backends in it

edgararuiz commented:
Hi @imanuelcostigan, thank you.  This translation is primarily meant to work with the 
[odbc](https://github.com/rstats-db/odbc) package

hadley commented on R/db-mssql.R:
```diff
@@ -0,0 +1,152 @@
+  info <- DBI::dbGetInfo(x)
+  host <- if (info$host == "") "localhost" else info$host
+
+  paste0("Microsoft SQL Server ", info$serverVersion, " [", info$user, "@",
```
Can you please follow the tidyverse indenting rules, i.e. http://style.tidyverse.org/syntax.html#long-lines ?

hadley commented on R/db-mssql.R:
```diff
@@ -0,0 +1,152 @@
+  out$select <- build_sql(
+    "SELECT ",
+
+    if (!is.null(limit) && !identical(limit, Inf)) {
```
It would be useful to have a comment here describing where SQL server deviates from SQL-92

hadley commented on R/db-mssql.R:
```diff
@@ -0,0 +1,152 @@
+
+
+#' @export
+`sql_subquery.Microsoft SQL Server` <- function(con, from, name = dbplyr:::unique_name(), ...) {
```
Isn't this the same as `sql_subquery.DBIConnection`?  You don't need to override methods that are the same as for DBIConnection.

hadley commented on R/db-mssql.R:
```diff
@@ -0,0 +1,152 @@
+      as.integer  = function(x) build_sql("CAST(", x, " AS INT)"),
+      as.logical = function(x) build_sql("CAST(", x, " AS BOOLEAN)"),
+      as.character  = function(x) build_sql("CAST(", x, " AS VARCHAR(MAX))"),
+      as.date  = function(x) build_sql("CAST(", x, " AS DATE)"),
```
There's no `as.date()` in base R?

hadley commented on R/db-mssql.R:
```diff
@@ -0,0 +1,152 @@
+      xor = function(x, y) build_sql(x, " ^ ", y),
+      or = function(x, y) build_sql(x, " or ", y),
+      and = function(x, y) build_sql(x, " and ", y),
+      pmin = function(...) build_sql_if_compare(..., con = con, compare = "<="),
```
Did you forget to check in this function?

hadley commented on R/db-mssql.R:
```diff
@@ -0,0 +1,152 @@
+
+
+#' @export
+`sql_join.Microsoft SQL Server` <- function(con, x, y, vars, type = "inner", by = NULL, ...) {
```
I think you can delete this too

edgararuiz commented:
Still working on the simulate_mssql tests,  not sure if they're required to pass this PR.

hadley commented on R/simulate.r:
```diff
@@ -32,3 +32,26 @@ simulate_postgres <- function() {
+
+# odbc connections --------------------------------------------------------
+
+setClass(
```
Can you please just simulate like the functions above?

edgararuiz replied:
Sure, quick question, do want me to use a single use function called `simulate_mssql`, or use the same code pattern as the other connections?:
```
simulate_postgres <- function() {
  structure(
    list(),
    class = c("PostgreSQLConnection", "DBIConnection")
  )
}
```

hadley replied:
New function - you'll need to mimic the classes that odbc generates

edgararuiz replied:
Got it, thanks

hadley commented on R/db-mssql.R:
```diff
@@ -0,0 +1,111 @@
+      as.character  = function(x) build_sql("CAST(", x, " AS VARCHAR(MAX))"),
+      as.Date  = function(x) build_sql("CAST(", x, " AS DATE)"),
+      paste0 = function(...) build_sql("CONCAT", list(...)),
+      or = function(x, y) build_sql(x, " or ", y),
```
Aren't these the default translations?

edgararuiz replied:
OR and AND were the only ones that passed the tests w/o translation, I removed them

hadley commented on tests/testthat/test-mssql.r:
```diff
@@ -0,0 +1,46 @@
+
+# Initialize simulated DB connection -------------------------------------
+
+mssql_connection <- dbplyr:::simulate_odbc("Microsoft SQL Server")
```
Can you please copy and paste into the individual tests? It's a bit hacky but it makes each test easier to explore interactively in isolation.

edgararuiz commented:
Hi @hadley , the latest commit passed two of the three Travis CI jobs.  It looks like the previously merged PR (#7) had the same issue, so I don't think it's due to something that needs to be fixed in this PR.  The error message mentions a mismatched dplyr versions inside the job's environment.  

Regarding the updates, I removed all of the code I previously used to simulate a MS SQL connection, and am using your new tbl_lazy method.  I also added a new test to confirm that `head()` translates to TOP

hadley commented on tests/testthat/test-mssql.r:
```diff
@@ -0,0 +1,104 @@
+  df_mssql <- tbl_lazy(df, src = simulate_mssql())
+  expect_equivalent(show_query(head(df_mssql)),
+                    sql("SELECT  TOP 6 *
+FROM `df`"))
```
Can you please use `\n` instead of the literal line break?

hadley commented on tests/testthat/test-mssql.r:
```diff
@@ -0,0 +1,104 @@
+                    sql("STDEV((`field_name`))"))
+})
+test_that("var aggregate functions translate to MSSQL VAR ", {
+  expect_equivalent(translate_sql(var(field_name),
```
Can you please use tidyverse style indenting? http://style.tidyverse.org/syntax.html#long-lines

hadley commented on NEWS.md:
```diff
@@ -150,6 +150,8 @@
   should make it easier to provide customised SQL for window functions
   (#2051, #2126).
+  
+*  SQL translation for Microsoft SQL Server
```
Can you please add `(@edgaruiz)`?

hadley commented:
Thanks!

edgararuiz commented:
Thank you for all the guidance!
